sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala

```diff
@@ -71,53 +71,61 @@ case class InlineCTE(
    * @param plan The plan to collect the CTEs from
    * @param cteMap A mutable map that accumulates the CTEs and their reference information by CTE
    *               ids.
-   * @param collectCTERefs A function to collect CTE references so that the caller side can do some
-   *                       bookkeeping work.
+   * @param outerCTEId While collecting the map we use this optional CTE id to identify the
+   *                   current outer CTE.
    */
   private def buildCTEMap(
       plan: LogicalPlan,
       cteMap: mutable.Map[Long, CTEReferenceInfo],
-      collectCTERefs: CTERelationRef => Unit = _ => ()): Unit = {
+      outerCTEId: Option[Long] = None): Unit = {
     plan match {
       case WithCTE(child, cteDefs) =>
+        var visit = true
```

Comment:
shall we more explicit about duplicated `WithCTE`? e.g.
```
val isDuplicated = cteDefs.forall(d => cteMap.contains(d.id))
```

Reply:
Yeah, sure. Fixed in https://github.com/apache/spark/pull/49379/commits/c62c6c4bf2ec4591fee1c5129315a2d9878159a9.